src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsImpl.hpp

```diff
@@ -0,0 +1,1447 @@
+        unit_interface_normal_vector,
+    const tnsr::iaa<DataVector, VolumeDim, Frame::Inertial>&
+        four_index_constraint,
+    const tnsr::iaa<DataVector, VolumeDim, Frame::Inertial>&
+        char_projected_rhs_dt_u_zero,
+    const std::array<DataVector, 4>& char_speeds) noexcept {
+  ASSERT(get_size(get<0, 0, 0>(*bc_dt_u_zero)) ==
+             get_size(get<0>(unit_interface_normal_vector)),
+         "Size of input variables and temporary memory do not match.");
+
+  for (size_t a = 0; a <= VolumeDim; ++a) {
+    for (size_t b = a; b <= VolumeDim; ++b) {
+      for (size_t i = 0; i < VolumeDim; ++i) {
+        bc_dt_u_zero->get(i, a, b) += char_projected_rhs_dt_u_zero.get(i, a, b);
+      }
+      // Lets say this term is T2_{kab} := - n_l N^l n^j C_{jkab}.
+      // But we store C4_{iab} = LeviCivita^{ijk} dphi_{jkab},
+      // which means  C_{jkab} = LeviCivita^{ijk} C4_{iab}
+      // where C4 is `four_index_constraint`.
+      // therefore, T2_{iab} =  char_speed<VZero> n^j C_{jiab}
+      // (since char_speed<VZero> = - n_l N^l), and therefore:
+      // T2_{iab} = char_speed<VZero> n^k LeviCivita^{ijk} C4_{jab}.
+      // Let LeviCivitaIterator be indexed by
+      // it[0] <--> i,
+      // it[1] <--> j,
+      // it[2] <--> k, then
+      // T2_{it[0], ab} += char_speed<VZero> n^it[2] it.sign() C4_{it[1], ab};
+      for (LeviCivitaIterator<VolumeDim> it; it; ++it) {
+        bc_dt_u_zero->get(it[0], a, b) +=
+            it.sign() * char_speeds.at(1) *
+            unit_interface_normal_vector.get(it[2] + 1) *
+            four_index_constraint.get(it[1], a, b);
+      }
+    }
+  }
+  return *bc_dt_u_zero;
+}
+
+template <typename ReturnType, size_t VolumeDim>
+ReturnType
+set_dt_u_zero<ReturnType, VolumeDim>::apply_dirichlet_constraint_preserving(
+    const gsl::not_null<ReturnType*> bc_dt_u_zero,
+    const tnsr::A<DataVector, VolumeDim, Frame::Inertial>&
+        unit_interface_normal_vector,
+    const tnsr::i<DataVector, VolumeDim, Frame::Inertial>&
+        unit_interface_normal_one_form,
+    const tnsr::A<DataVector, VolumeDim, Frame::Inertial>&
+        spacetime_unit_normal_vector,
+    const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, VolumeDim, Frame::Inertial>& shift,
+    const tnsr::II<DataVector, VolumeDim, Frame::Inertial>&
+        inverse_spatial_metric,
+    const tnsr::aa<DataVector, VolumeDim, Frame::Inertial>& pi,
+    const tnsr::iaa<DataVector, VolumeDim, Frame::Inertial>& phi,
+    const tnsr::iaa<DataVector, VolumeDim, Frame::Inertial>& d_psi,
+    const tnsr::iaa<DataVector, VolumeDim, Frame::Inertial>& d_pi,
+    const tnsr::ijaa<DataVector, VolumeDim, Frame::Inertial>& d_phi) noexcept {
+  ASSERT(get_size(get<0, 0, 0>(*bc_dt_u_zero)) == get_size(get(lapse)),
+         "Size of input variables and temporary memory do not match.");
+  for (size_t a = 0; a <= VolumeDim; ++a) {
+    for (size_t b = 0; b <= VolumeDim; ++b) {
+      // For a chosen (a, b):
```

Comment:
comments like this should be doxygen comments so we can check the typesetting etc.